### PR TITLE
подбор параметров

### PR DESCRIPTION
**Pull request by JuliaBars**
Оценка производительности модели через KFold кросс-валидацию при сочетании  параметров: 

Структура слоев:  
- [512, 512, 512]  
- [1024, 512, 1024]  
- [512, 256, 512]  

Набор dropout:   
- {0.1, 0.3, 0.5}

Повторить эксперименты 15 и 17 с измененными параметрами EarlyStopping
- ('min_delta': 0.005, 'patience': 20`)

Выводы:
1. наилучший результат показала модель с [1024, 512, 1024] и `dropuot` 0.1, но     превосходит другие незначительно, возможно погрешность.
2. применение регуляризации сильно снижает показатели ~ 0.15-0.17  по mean_f1_score_micro
3. регулирование параметров EarlyStopping никаких особо значимых различий не принесло.

Добавлен файлик limit_data_by_terms, как вариант выполнения задачи по выбору 1500 самых редких термов.

### CODE REVIEW

File: src/logger.py [hunk 1]
```diff
@@ -27,3 +44,52 @@ def log_to_json(
     with open(log_file, 'w') as f:
         json.dump(data, f, indent=4, ensure_ascii=False)
         print(f'Файл был {result}')
+
+
+def log_to_table(log_file: Path) -> pd.DataFrame:
```
Comment (Yo1k) [inline src/logger.py:49]: У нас может быть несколько логов (`log_to_json` мы можем использовать для сохранения экспериментов по оценки производительности моделей (в файле типа `eval.json`), а можем и для логирования информации по датасетам и их параметрам их формирования). Данная функция работает только с определенным логом, поэтому это хорошо в названии отражать: `eval_log_to_json`.

File: notebooks/jo/model_creator_jo.ipynb [hunk 2]
```diff
@@ -0,0 +1,1123 @@
...
+    "log_to_table(DIR_ROOT / 'experiments/eval.json')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Выводы\n",
+    "\n",
+    "1. Наилучший результат показала модель №15 с [1024, 512, 1024] и `dropuot` 0.1, но  \n",
```
Comment (Yo1k) [inline notebooks/jo/model_creator_jo.ipynb:1090]: __На заметку__
Производительность моделей для разных гиперпараметров ты оцениваешь по метрике `f1_score_micro`. Да, есть таблица, но чтобы проверить твой вывод нужно долго всматриваться в таблицу. Чтобы упростить проверку твоего вывода, ты могла сделать сортировку по убыванию для поля  `f1_score_micro` в своей таблице. Так и тебе было бы легче оценивать данные и читателям проверять твои выводы. См. метод [`sort_values`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html)

**Review by Yo1k — CHANGES_REQUESTED** [review]

File: src/model_utils.py [hunk 3]
```diff
@@ -152,12 +152,14 @@ def evaluate_model(
         n_repeats=n_repeats,
         random_state=random_state,
     )
+    history_all_epochs = []
```
[the remark this answers is not in the document]
Comment (JuliaBars) [reply src/model_utils.py:155]: DONE

File: src/model_utils.py [hunk 4]
```diff
@@ -166,7 +168,8 @@ def evaluate_model(
             )
         # Очистка объектов.
         proxy_model.clear()
-    return results
+
+    return results, history_all_epochs
```
[the remark this answers is not in the document]
Comment (JuliaBars) [reply src/model_utils.py:172]: DONE

File: src/logger.py [hunk 5]
```diff
@@ -1,5 +1,22 @@
 import json
 from pathlib import Path
+import pandas as pd
+from statistics import mean
+
+import tensorflow as tf
+
+
+def mean_callback_epoch(
+    history_all_epochs: tf.keras.callbacks.History
```
[the remark this answers is not in the document]
Comment (JuliaBars) [reply src/logger.py:10]: DONE

[the remark this answers is not in the document]
Comment (JuliaBars) [reply src/logger.py:10]: DONE

File: src/logger.py [hunk 6]
```diff
@@ -1,5 +1,22 @@
 import json
 from pathlib import Path
+import pandas as pd
+from statistics import mean
+
+import tensorflow as tf
+
+
+def mean_callback_epoch(
```
[the remark this answers is not in the document]
Comment (JuliaBars) [reply src/logger.py:9]: DONE

File: src/logger.py [hunk 7]
```diff
@@ -1,5 +1,22 @@
...
+    history_all_epochs: tf.keras.callbacks.History
+) -> float:
+    '''
+    Расчитываем среднюю `callback` эпоху для всех `KFold`.
```
[the remark this answers is not in the document]
Comment (JuliaBars) [reply src/logger.py:13]: DONE

File: src/logger.py [hunk 8]
```diff
@@ -1,5 +1,22 @@
...
+    '''
+    callback_epochs: list[int] = []
```
[the remark this answers is not in the document]
Comment (JuliaBars) [reply src/logger.py:15]: DONE

File: src/logger.py (hunk 1 above)
Comment (JuliaBars) [reply src/logger.py:49]: DONE

File: notebooks/jo/model_creator_jo.ipynb (hunk 2 above)
Comment (JuliaBars) [reply notebooks/jo/model_creator_jo.ipynb:1090]: DONE

**Review by Yo1k — APPROVED** [review]
